Sources/SwiftKuery/QueryResult.swift

```diff
@@ -80,38 +80,72 @@ public enum QueryResult {
         }
     }
     
-    /// Data received from the query execution represented as an array of dictionaries.
+    /// Passes data received from the query execution represented as an array of dictionaries to a user defined completion handler.
     /// Each entry in the array repersents a row in the result with column titles as the keys.
     /// In case there are columns with the same title, we add indices to the keys: for example if
     /// the result contains three columns named 'a', the dictionary will contain the keys: a, a.1 and a.2.
     /// This function will consume the entire set of results and then call done against the result set which will allow the underlying connection to be reused.
     /// When asRows is called a user does not need to explicitly invoke the done method on the result set.
-    public var asRows: [[String:Any?]]? {
+    public func asRows(onCompletion: @escaping (([[String:Any?]]?, Error?)) -> ()) {
         switch self {
         case .resultSet(let resultSet):
-            let z = resultSet.rows.map { zip(resultSet.titles, $0) }
-            let arrayOfDictionaries = z.map { row -> [String:Any?] in
-                var dictionary = [String:Any?]()
-                for (title, value) in row {
-                    if dictionary[title] == nil {
-                        dictionary[title] = value
-                    }
-                    else {
-                        // In case we have a collision on key names.
-                        var i = 1
-                        while (dictionary[title + ".\(i)"] != nil) {
-                            i = i + 1
-                        }
-                        dictionary[title + ".\(i)"] = value
+            resultSet.getColumnTitles() { titles, error in
+                guard let columnTitles = titles else {
+                    guard let error = error else {
+                        return onCompletion((nil, QueryError.noResult("Unable to retrieve Column titles")))
                     }
+                    return onCompletion((nil, QueryError.noResult(error.localizedDescription)))
                 }
-                return dictionary
+                return self.processRows(resultSet, titles: columnTitles, result: nil, onCompletion: onCompletion)
             }
-            // As this function consumes the entire result set we can explicitly close the result set allowing the release of the underlying connection.
-            resultSet.done()
-            return arrayOfDictionaries
+            return
+        case .successNoData:
+            return onCompletion((nil, QueryError.noResult("Operation was succesful but no results returned.")))
+        case .error(let error):
+            return onCompletion((nil, QueryError.noResult("Result is an error: \(error.localizedDescription)")))
         default:
```

Comment:
Could this be .success? It would make it clearer which errors correspond to which situation.

Reply:
It can and updated.